### PR TITLE
Implement execution SLA checks for jobs in containerized executions.

### PR DESCRIPTION
**Pull request by ypadron-in**
This PR implements execution SLA checks for jobs in containerized executions.
It also makes some refactoring to the executions SLA feature.

Tested locally in solo server, testing clusters and Kubernes pods. 
Verified that both Flow and Job SLA checks are setup and alert/kill actions are taken upon SLA misses.

Follow up: work on adding tests for the feature overall which do not exist today.

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/executor/IFlowRunnerManager.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package azkaban.executor;
+
+import azkaban.utils.FileIOUtils.LogData;
+
+/**
+ * Defines operations common to all flow execution controllers.
+ */
+public interface IFlowRunnerManager {
```
Comment (djaiswal83) [inline azkaban-common/src/main/java/azkaban/executor/IFlowRunnerManager.java:24]: This interface is not needed as FlowContainer does not implement half of the APIs and does not require execId at all.

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java [hunk 2]
```diff
@@ -114,7 +117,7 @@
  *  RUNNING from FlowRunner. The rest of the state machine is handled by FlowRunner.
  */
 @Singleton
-public class FlowContainer implements IMBeanRegistrable, EventListener<Event> {
+public class FlowContainer implements IFlowRunnerManager, IMBeanRegistrable, EventListener<Event> {
```
Comment (djaiswal83) [inline azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:120]: Please remove the IFlowRunnerManager

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java [hunk 3]
```diff
@@ -509,13 +525,26 @@ public void cancelFlow(final int execId, final String user)
...
+  public void cancelJobBySLA(final int execId, final String jobId)
+      throws ExecutorManagerException {
+    if (this.flowRunner == null) {
+      throw new ExecutorManagerException("Execution " + execId + " is not running.");
+    }
+
+    this.flowRunner.getExecutableFlow().setModifiedBy("SLA");
+
+    for (final JobRunner jobRunner : this.flowRunner.getActiveJobRunners()) {
+      if (jobRunner.getJobId().equals(jobId)) {
```
Comment (djaiswal83) [inline azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:538]: Please put negative case here and put the positive case after that.
if (!jobRunner.getJobId().equals(jobId)) { 
  continue};
}
<rest of the code>

**Review by djaiswal83 — CHANGES_REQUESTED** [review]
Thanks for implementing the missing feature in FlowContainer.
The idea of the interface is not very neat here and not required.
Also, passing a top level object such as FlowRunnerManager to Kill*Action classes does not seem to be the right way to do it. 
Thanks for cleaning up the unused APIs.

File: azkaban-common/src/main/java/azkaban/executor/IFlowRunnerManager.java (hunk 1 above)
Comment (ypadron-in) [reply azkaban-common/src/main/java/azkaban/executor/IFlowRunnerManager.java:24]: Could you please elaborate on why the interface isn't needed? 
-As I see it FlowContainer and FlowRunnerManager have the same main responsibility: to control a flow execution.
-Trigger action objects from `KillExecutionAction` and `KillJobAction` depend on such flow execution controllers for their functionality. Without the interface then we need to go back to if-else of current DispatchMethod and using  `SERVICE_PROVIDER.getInstance` of Guice which I believe is an anti-pattern and not so much extensible/maintainable code.

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java (hunk 3 above)
Comment (ypadron-in) [reply azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:538]: Do you mean like this?
```
{ // start for loop
  if (!jobRunner.getJobId().equals(jobId)) {
    continue
  };
  logger.info("Killing job " + jobId + " in execution " + execId + " by SLA");
  jobRunner.killBySLA();
  break;
} // end for
```
I understand that asking for the negative case is better overall but given the simplicity of the logic (log and call kill action) existing code looks more readable and simple to me than above block. I can definitely make the change if you feel strongly about it though.

**Review by jakhani — APPROVED** [review]
Can you please check if you can add unit tests for SLA kill? Thank you for working on this PR.

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java (hunk 3 above)
Comment (djaiswal83) [reply azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:538]: This code may become larger if further development happens here for some reason in the future. It is better to keep it the way suggested for such cases.

**Review by djaiswal83 — CHANGES_REQUESTED** [review]
I resolved my interface related comments. I just have one comment, please address that. Overall the code looks much cleaner due to this PR. Thanks!

**Review by djaiswal83 — APPROVED** [review]
Thanks for addressing the comments. The code looks good.

**Comment by ypadron-in** [comment]
> Can you please check if you can add unit tests for SLA kill? Thank you for working on this PR.

I'll add tests in a follow up PR

**Pull request merged**